test/functional/sqlsrv/sqlsrv_ae_datetimes_as_strings.phpt

```diff
@@ -0,0 +1,627 @@
+// 'Y-m-d H:i:s.u P', which is the format returned when $returnDatesAsStrings 
+// is true, or by the date_format() calls in FetchDatesAndOrTimes when a PHP 
+// DateTime object is retrieved (i.e. when $returnDatesAsStrings is false),
+// unless $datetimetype is date or time. In those cases:
+// If $datetimetype is date and $returnDatesAsStrings is false:
+//   The date is as expected, the time defaults to 00:00:00.0000
+// If $datetimetype is time and $returnStrings is false:
+//   The date defaults to the current date, the time is as expected. 
+// If $datetimetype is date and $returnStrings is true:
+//   $retrievedDateTime is only a date.
+// If $datetimetype is 'time' and $returnStrings is true:
+//   $retrievedDateTime is only a time.
+function CompareDateTime($datetimetype, $returnStrings, &$expectedDateTime, $retrievedDateTime)
+{
+    $expected_date_time = array();
+
+    // Split each element of the testing date/time into
+    // [0]:date, [1]:time, and possibly [2]:timezone offset 
+    for ($i = 0; $i < sizeof($expectedDateTime); ++$i) {
+        $expected_date_time[] = explode(" ", $expectedDateTime[$i]);
+    }
+    
+    // If $retrievedDateTime is a string of format 'Y-m-d H:i:s.u' or 'Y-m-d H:i:s.u P',
+    // split it into [0]:date, [1]:time, and possibly [2]:timezone offset
+    $retrieved_date_time = explode(" ", $retrievedDateTime);
+   
+    if ($returnStrings == true) {
+        switch ($datetimetype) {
+            case 'date':
+                // Direct comparison of retrieved date and expected date
+                if ($retrievedDateTime != $expected_date_time[0][0]) {
+                    fatalError("Dates do not match!");
+                }                    
+                break;
+            case 'time': 
+                // Compare SQL time with expected time. The expected time was input
+                // with an accuracy of microseconds and the SQL Servertime type has
+                // accuracy to 100 ns, so times are returned with an extra zero. For
+                // comparison the zero is appended to the times in expected_time_date.
+                if ($retrievedDateTime != $expected_date_time[0][1]."0" and
+                    $retrievedDateTime != $expected_date_time[1][1]."0" and
+                    $retrievedDateTime != $expected_date_time[2][1]."0" and
+                    $retrievedDateTime != $expected_date_time[3][1]."0" ) {
+                        fatalError("Times do not match!");
+                }
+                break;
+            case 'datetime':
+                // Compare retrieved SQL datetime with expected date/time.
+                // SQL Server's datetime type is accurate to milliseconds and
+                // the expected time is accurate to microseconds, so append
+                // three zeroes to the retrieved time for comparison.
+                if ($retrievedDateTime."000" != $expectedDateTime[0] and
```

Comment:
not sure why this passed. `$expectedDateTime[0]` is `$year."-".$month."-".$day." ".$hour.":".$minute.":".$second.".".$frac.$frac2`, and `$frac2` is `'9876'`.

When the input is first inserted, the input is `'2002-01-31 23:59:29.049876'`. The actual data inserted however should be `'2002-01-31 23:59:29.049'` since datetime only goes up to a precision of 0.001s.

When the data is retrieved, the `$retrievedDateTime` is `'2002-01-31 23:59:29.049'`, and you append `"000"` to it which makes it `'2002-01-31 23:59:29.049000'`, however, that is obviously not the same as `'2002-01-31 23:59:29.049876'`. So this condition should've evaluated to false.